src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/CompletionListOptimizer.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using AliasedVSCommitCharacters = Microsoft.VisualStudio.LanguageServer.Protocol.SumType<string[], Microsoft.VisualStudio.LanguageServer.Protocol.VSInternalCommitCharacter[]>;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Completion
+{
+    internal static class CompletionListOptimizer
+    {
+        public static VSInternalCompletionList Optimize(VSInternalCompletionList completionList, VSInternalCompletionSetting? completionCapability)
+        {
+            if (completionCapability is not null)
+            {
+                completionList = OptimizeCommitCharacters(completionList, completionCapability);
+                completionList = OptimizeData(completionList, completionCapability);
+            }
+
+            // We wrap the pre-existing completion list with an optimized completion list to better control serialization/deserialization
+            var optimizedCompletionList = new OptimizedVSCompletionList(completionList);
+            return optimizedCompletionList;
+        }
+
+        private static VSInternalCompletionList OptimizeCommitCharacters(VSInternalCompletionList completionList, VSInternalCompletionSetting completionCapability)
+        {
+            var completionListCapability = completionCapability.CompletionList;
+            if (completionListCapability?.CommitCharacters != true)
+            {
+                return completionList;
+            }
+
+            // The commit characters capability is a VS capability with how we utilize it, therefore we want to promote onto the VS list.
+            completionList = PromoteVSCommonCommitCharactersOntoList(completionList);
+            return completionList;
+        }
+
+        private static VSInternalCompletionList OptimizeData(VSInternalCompletionList completionList, VSInternalCompletionSetting completionCapability)
+        {
+            var completionListCapability = completionCapability.CompletionList;
+            if (completionListCapability?.Data != true)
+            {
+                return completionList;
+            }
+
+            completionList = PromotedDataOntoList(completionList);
+
+            return completionList;
+        }
```

Comment:
nit: 
```suggestion
        }
        
```

Reply:
Nit on your nit 😛 (even though I know it was really githubs fault):
```suggestion
        }

```